File: src/main/scala/vm/vagrant/configuration/builder/VagrantVmConfigBuilder.scala
package vm.vagrant.configuration.builder

import java.io.File

import scala.collection.immutable.List
import java.net.MalformedURLException
import java.net.URL
import java.util.UUID

import vm.vagrant.configuration._
import vm.vagrant.configuration.ChecksumType.ChecksumType
import vm.vagrant.configuration.builder.util.VagrantBuilderException


object VagrantVmConfigBuilder {
  def create = new VagrantVmConfigBuilder
}

class VagrantVmConfigBuilder() {
  private var name: String = UUID.randomUUID.toString
  private var hostName: String = _
  private var boxName: String = _
  private var boxUrl: URL = _
  private var vagrantProvisionerConfigs: List[VagrantProvisionerConfig] = _
  private var vagrantNetworkConfigs: List[VagrantNetworkConfig] = _
  private var vagrantSyncedFolderConfigs: List[VagrantSyncedFolderConfig] = _
  private var guiMode: Boolean = false
  private var bootTimeout: Int = 300
  private var boxCheckUpdate: Boolean = true
  private var boxDownloadChecksum: String = _
  private var boxDownloadChecksumType: ChecksumType = _
  private var boxDownloadClientCert: File = _
  private var boxDownloadCaCert: File = _
  private var boxDownloadCaPath: File = _
  private var boxDownloadInsecure: Boolean = false
  private var boxDownloadLocationTrusted: Boolean = false
  private var boxVersion: String = _
  private var communicator: String = _
  private var gracefulHaltTimeout: Int = 60
  private var guest: String = _
  private var postUpMessage: String = _
  private var usablePortRange: String = "2200..2250"
  private var provider: VagrantProviderConfig = _


  def withName(name: String): VagrantVmConfigBuilder = {
    this.name = name
    this
  }

  def withHostName(hostName: String): VagrantVmConfigBuilder = {
    this.hostName = hostName
    this
  }

  def withBoxName(boxName: String): VagrantVmConfigBuilder = {
    this.boxName = boxName
    this
  }

  def withBoxUrl(boxUrl: URL): VagrantVmConfigBuilder = {
    this.boxUrl = boxUrl
    this
  }

  def withVagrantProvisionerConfig(vagrantProvisionerConfig: VagrantProvisionerConfig): VagrantVmConfigBuilder = {
    this.vagrantProvisionerConfigs ::= vagrantProvisionerConfig
    this
  }

  def withVagrantNetworkConfig(vagrantNetworkConfig: VagrantNetworkConfig): VagrantVmConfigBuilder = {
    this.vagrantNetworkConfigs ::= vagrantNetworkConfig
    this
  }

  def withVagrantSyncedFoldersConfig(vagrantSyncedFoldersConfig: VagrantSyncedFolderConfig): VagrantVmConfigBuilder = {
    this.vagrantSyncedFolderConfigs ::= vagrantSyncedFoldersConfig
    this
  }

  def withGuiMode(guiMode: Boolean): VagrantVmConfigBuilder = {
    this.guiMode = guiMode
    this
  }

  def withBootTimeout(bootTimeout: Int): VagrantVmConfigBuilder = {
    this.bootTimeout = bootTimeout
    this
  }

  def withBoxCheckUpdate(boxCheckUpdate: Boolean): VagrantVmConfigBuilder = {
    this.boxCheckUpdate = boxCheckUpdate
    this
  }

  def withBoxDownloadChecksum(boxDownloadChecksum: String): VagrantVmConfigBuilder = {
    this.boxDownloadChecksum = boxDownloadChecksum
    this
  }

  def withBoxDownloadChecksumType(boxDownloadChecksumType: ChecksumType): VagrantVmConfigBuilder = {
    this.boxDownloadChecksumType = boxDownloadChecksumType
    this
  }

  def withBoxDownloadClientCert(boxDownloadClientCert: File): VagrantVmConfigBuilder = {
    this.boxDownloadClientCert = boxDownloadClientCert
    this
  }

  def withBoxDownloadCaCert(boxDownloadCaCert: File): VagrantVmConfigBuilder = {
    this.boxDownloadCaCert = boxDownloadCaCert
    this
  }

  def withBoxDownloadCaPath(boxDownloadCaPath: File): VagrantVmConfigBuilder = {
    this.boxDownloadCaPath = boxDownloadCaPath
    this
  }

  def withBoxDownloadInsecure(boxDownloadInsecure: Boolean): VagrantVmConfigBuilder = {
    this.boxDownloadInsecure = boxDownloadInsecure
    this
  }

  def withBoxDownloadLocationTrusted(boxDownloadLocationTrusted: Boolean): VagrantVmConfigBuilder = {
    this.boxDownloadLocationTrusted = boxDownloadLocationTrusted
    this
  }

  def withBoxVersion(boxVersion: String): VagrantVmConfigBuilder = {
    this.boxVersion = boxVersion
    this
  }

  def withCommunicator(communicator: String): VagrantVmConfigBuilder = {
    this.communicator = communicator
    this
  }

  def withGracefulHaltTimeout(gracefulHaltTimeout: Int): VagrantVmConfigBuilder = {
    this.gracefulHaltTimeout = gracefulHaltTimeout
    this
  }

  def withGuest(guest: String): VagrantVmConfigBuilder = {
    this.guest = guest
    this
  }

  def withPostUpMessage(postUpMessage: String): VagrantVmConfigBuilder = {
    this.postUpMessage = postUpMessage
    this
  }

  def withUsablePortRange(usablePortRange: String): VagrantVmConfigBuilder = {
    this.usablePortRange = usablePortRange
    this
  }

  def withProvider(provider: VagrantProviderConfig): VagrantVmConfigBuilder = {
    this.provider = provider
    this
  }

  /*
   * SSP - Part !!!!!!!!!!!!!!!!!!!!!!!!!
   */

  def withSspDevBox: VagrantVmConfigBuilder = {
    this.boxName = "sds-devbox"
    try
      this.boxUrl = new URL("https://sds.ssp-europe.eu/api/v4/public/shares/downloads/YgegiC6hF3mtmSFoQ6IaLfF5maaIRGYm/5rogL8eq-fr2ZYhO6jW1XMxbbZxlaDw0L5gW7lj4v_mqMIssU1Ie79BRj9hHVDpBhcOU3YWWbELPUcYCOKAu47t4_y_8Mv-rmEnJLYbC7DDVz51is9AOd9VE5zdeU5ym4_HcGnX4TBLYCTwn3RiIcDo_EwCPnJ0h_RiKLG8_d26dgTohAl_KMYHQ1e4f18df7e302f02")
    catch {
      case e: MalformedURLException =>
        throw new RuntimeException(e)
    }
    this
  }

  /*
   * SSP - Part !!!!!!!!!!!!!!!!!!!!!!!!!
   */

  def build: VagrantVmConfig = {
    if (boxName == null) throw new VagrantBuilderException("No boxName defined")
    new VagrantVmConfig(name = name,
      hostName = hostName,
      boxName = boxName,
      boxUrl = boxUrl,
      vagrantProvisionerConfigs = vagrantProvisionerConfigs,
      vagrantNetworkConfigs = vagrantNetworkConfigs,
      vagrantSyncedFolderConfigs = vagrantSyncedFolderConfigs,
      guiMode = guiMode,
      bootTimeout = bootTimeout,
      boxCheckUpdate = boxCheckUpdate,
      boxDownloadChecksum = boxDownloadChecksum,
      boxDownloadChecksumType = boxDownloadChecksumType,
      boxDownloadClientCert = boxDownloadClientCert,
      boxDownloadCaCert = boxDownloadCaCert,
      boxDownloadCaPath = boxDownloadCaPath,
      boxDownloadInsecure = boxDownloadInsecure,
      boxDownloadLocationTrusted = boxDownloadLocationTrusted,
      boxVersion = boxVersion,
      communicator = communicator,
      gracefulHaltTimeout = gracefulHaltTimeout,
      guest = guest,
      postUpMessage = postUpMessage,
      usablePortRange = usablePortRange,
      provider = provider)
  }
}

File: src/main/java/vm/vagrant/configuration/builder/VagrantVmConfigBuilder.java
package vm.vagrant.configuration.builder;

import vm.vagrant.configuration.*;
import vm.vagrant.configuration.builder.util.VagrantBuilderException;
import vm.vagrant.util.ChecksumType;

import java.io.File;
import java.io.Serializable;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class VagrantVmConfigBuilder implements Serializable {
    private String name;
    private String hostName;
    private String boxName;
    private URL boxUrl;
    private List<VagrantProvisionerConfig> vagrantProvisionerConfigs;
    private List<VagrantNetworkConfig> vagrantNetworkConfigs;
    private List<VagrantSyncedFolderConfig> vagrantSyncedFolderConfigs;
    private Integer bootTimeout;
    private Boolean boxCheckUpdate;
    private String boxDownloadChecksum;
    private ChecksumType boxDownloadChecksumType;
    private File boxDownloadClientCert;
    private File boxDownloadCaCert;
    private File boxDownloadCaPath;
    private Boolean boxDownloadInsecure;
    private Boolean boxDownloadLocationTrusted;
    private String boxVersion;
    private String communicator;
    private Integer gracefulHaltTimeout;
    private String guest;
    private String postUpMessage;
    private String usablePortRange;
    private VagrantProviderConfig provider;

    public VagrantVmConfigBuilder() {
        vagrantProvisionerConfigs = new ArrayList<VagrantProvisionerConfig>();
        vagrantNetworkConfigs = new ArrayList<VagrantNetworkConfig>();
        vagrantSyncedFolderConfigs = new ArrayList<VagrantSyncedFolderConfig>();
        name = UUID.randomUUID().toString();
        bootTimeout = 300;
        boxCheckUpdate = true;
        boxDownloadInsecure = false;
        boxDownloadLocationTrusted = false;
        gracefulHaltTimeout = 60;
        usablePortRange = "2200..3200";
    }

    public static VagrantVmConfigBuilder create() {
        return new VagrantVmConfigBuilder();
    }

    public VagrantVmConfigBuilder withName(String name) {
        this.name = name;
        return this;
    }

    public VagrantVmConfigBuilder withHostName(String hostName) {
        this.hostName = hostName;
        return this;
    }

    public VagrantVmConfigBuilder withBoxName(String boxName) {
        this.boxName = boxName;
        return this;
    }

    public VagrantVmConfigBuilder withBoxUrl(URL boxUrl) {
        this.boxUrl = boxUrl;
        return this;
    }

    public VagrantVmConfigBuilder withVagrantProvisionerConfig(VagrantProvisionerConfig vagrantProvisionerConfigs) {
        this.vagrantProvisionerConfigs.add(vagrantProvisionerConfigs);
        return this;
    }

    public VagrantVmConfigBuilder withVagrantNetworkConfig(VagrantNetworkConfig vagrantNetworkConfigs) {
        this.vagrantNetworkConfigs.add(vagrantNetworkConfigs);
        return this;
    }

    public VagrantVmConfigBuilder withVagrantSyncedFolderConfig(VagrantSyncedFolderConfig vagrantSyncedFolderConfigs) {
        this.vagrantSyncedFolderConfigs.add(vagrantSyncedFolderConfigs);
        return this;
    }

    public VagrantVmConfigBuilder withBootTimeout(Integer bootTimeout) {
        this.bootTimeout = bootTimeout;
        return this;
    }

    public VagrantVmConfigBuilder withBoxCheckUpdate(Boolean boxCheckUpdate) {
        this.boxCheckUpdate = boxCheckUpdate;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadChecksum(String boxDownloadChecksum) {
        this.boxDownloadChecksum = boxDownloadChecksum;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadChecksumType(ChecksumType boxDownloadChecksumType) {
        this.boxDownloadChecksumType = boxDownloadChecksumType;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadClientCert(File boxDownloadClientCert) {
        this.boxDownloadClientCert = boxDownloadClientCert;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadCaCert(File boxDownloadCaCert) {
        this.boxDownloadCaCert = boxDownloadCaCert;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadCaPath(File boxDownloadCaPath) {
        this.boxDownloadCaPath = boxDownloadCaPath;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadInsecure(Boolean boxDownloadInsecure) {
        this.boxDownloadInsecure = boxDownloadInsecure;
        return this;
    }

    public VagrantVmConfigBuilder withBoxDownloadLocationTrusted(Boolean boxDownloadLocationTrusted) {
        this.boxDownloadLocationTrusted = boxDownloadLocationTrusted;
        return this;
    }

    public VagrantVmConfigBuilder withBoxVersion(String boxVersion) {
        this.boxVersion = boxVersion;
        return this;
    }

    public VagrantVmConfigBuilder withCommunicator(String communicator) {
        this.communicator = communicator;
        return this;
    }

    public VagrantVmConfigBuilder withGracefulHaltTimeout(Integer gracefulHaltTimeout) {
        this.gracefulHaltTimeout = gracefulHaltTimeout;
        return this;
    }

    public VagrantVmConfigBuilder withGuest(String guest) {
        this.guest = guest;
        return this;
    }

    public VagrantVmConfigBuilder withPostUpMessage(String postUpMessage) {
        this.postUpMessage = postUpMessage;
        return this;
    }

    public VagrantVmConfigBuilder withUsablePortRange(String usablePortRange) {
        this.usablePortRange = usablePortRange;
        return this;
    }

    public VagrantVmConfigBuilder withProvider(VagrantProviderConfig provider) {
        this.provider = provider;
        return this;
    }

    public VagrantVmConfig build() {
        if (boxName == null && !boxName.isEmpty()) throw new VagrantBuilderException("No boxName defined");
        if (provider == null) throw new VagrantBuilderException("No provider defined");
        return new VagrantVmConfig(name, hostName, boxName, boxUrl, vagrantProvisionerConfigs, vagrantNetworkConfigs, vagrantSyncedFolderConfigs, bootTimeout, boxCheckUpdate, boxDownloadChecksum, boxDownloadChecksumType, boxDownloadClientCert, boxDownloadCaCert, boxDownloadCaPath, boxDownloadInsecure, boxDownloadLocationTrusted, boxVersion, communicator, gracefulHaltTimeout, guest, postUpMessage, usablePortRange, provider);
    }
}
